python/ray/widgets/util.py

```diff
@@ -59,3 +73,93 @@ def make_table_html_repr(
+    missing_message: Optional[str] = None,
+    outdated_message: Optional[str] = None,
+) -> Callable[[F], F]:
+    """Generate a decorator which checks for soft dependencies.
+
+    This decorator is meant to wrap _ipython_display_. If the dependency is not found,
+    or a version is specified here and the version of the package is older than the
+    specified version, the wrapped function is not executed and None is returned. If
+    the dependency is missing or the version is old, a log message is displayed.
+
+    Args:
+        *deps: Iterable of (dependency name, min version (optional))
+        missing_message: Message to log if missing package is found
+        outdated_message: Message to log if outdated package is found
+
+    Returns:
+        Wrapped function. Guaranteed to be safe to import soft dependencies specified
+        above.
+    """
+
+    def wrapper(func: F) -> F:
+        @wraps(func)
+        def wrapped(*args, **kwargs):
+            if _has_missing(*deps, message=missing_message) or _has_outdated(
+                *deps, message=outdated_message
+            ):
+                return None
+            return func(*args, **kwargs)
+
+        return wrapped
+
+    return wrapper
+
+
+def _has_missing(
+    *deps: Iterable[Union[str, Optional[str]]], message: Optional[str] = None
+):
+    missing = []
+    for (lib, _) in deps:
+        try:
+            importlib.import_module(lib)
+        except ImportError:
+            missing.append(lib)
+
+    if missing:
+        if not message:
+            message = f"Run `pip install {' '.join(missing)}` for rich notebook output."
+
+        # stacklevel=3: First level is this function, then ensure_notebook_deps, then
+        # the actual function affected.
+        logger.warning(f"Missing packages: {missing}. {message}", stacklevel=3)
+
+    return missing
+
+
+def _has_outdated(
+    *deps: Iterable[Union[str, Optional[str]]], message: Optional[str] = None
+):
+    outdated = []
+    for (lib, version) in deps:
+        try:
+            module = importlib.import_module(lib)
+            if version and Version(module.__version__) < Version(version):
+                outdated.append([lib, version, module.__version__])
+        except ImportError:
+            pass
+
+    if outdated:
+        outdated_strs = []
+        install_args = []
+        for lib, version, installed in outdated:
+            outdated_strs.append(f"{lib}=={installed} found, needs {lib}>={version}")
+            install_args.append(f"{lib}>={version}")
+
+        outdated_str = textwrap.indent("\n".join(outdated_strs), "  ")
+        install_str = " ".join(install_args)
+
+        if not message:
+            message = f"Run `pip install -U {install_str}` for rich notebook output."
+
+        # stacklevel=3: First level is this function, then ensure_notebook_deps, then
+        # the actual function affected.
+        logger.warning(f"Outdated packages:\n{outdated_str}\n{message}", stacklevel=3)
```

Comment:
Looks like this is causing issues for Python 3.7. `stacklevel` wasn't introduced until Python 3.8, so I'm getting this error:

```
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/IPython/core/formatters.py in __call__(self, obj)
    916             method = get_real_method(obj, self.print_method)
    917             if method is not None:
--> 918                 method()
    919                 return True
    920 

2 frames
/usr/lib/python3.7/logging/__init__.py in warning(self, msg, *args, **kwargs)
   1388         """
   1389         if self.isEnabledFor(WARNING):
-> 1390             self._log(WARNING, msg, args, **kwargs)
   1391 
   1392     def warn(self, msg, *args, **kwargs):

TypeError: _log() got an unexpected keyword argument 'stacklevel'
```